File: py/file_reader/reader.py
from os import path, scandir

from constants import ROOT
from .stack import Stack
from .tree_node import TreeNode


def create_dir_tree() -> TreeNode:
    file_stack = Stack[TreeNode]()
    root_node = TreeNode(ROOT, path.getsize(ROOT))

    file_stack.append(root_node)

    while not file_stack.is_empty():
        current_node = file_stack.pop()

        if not path.isdir(current_node.path):
            continue

        try:
            dir_list = scandir(current_node.path)
        except PermissionError:
            pass

        for file in dir_list:
            if file.is_symlink():
                continue

            new_node = TreeNode(file.path, file.stat().st_size)

            if file.is_dir():
                file_stack.append(new_node)

            current_node.append_child(new_node)

    return root_node

File: ts/fileSystem/reader.ts
import { readdir, lstat } from 'fs/promises';

import { ROOT } from '../constants/index.js';
import { allowedErrorCodes, isNodeError } from '../utils/Error.js';
import { Stack } from '../utils/Stack.js';
import { TreeNode } from './TreeNode.js';

// Scans only directories that user has permission to access
// Walk iteratively through system files
export async function createDirTree(): Promise<TreeNode> {
  const fileStack = new Stack<TreeNode>();
  const rootNode = new TreeNode(ROOT);

  fileStack.push(rootNode);

  while (!fileStack.isEmpty()) {
    const currentNode = fileStack.pop();
    try {
        const fileNameList = await readdir(currentNode.getAbsPath());
        const dirPath = currentNode.getAbsPath();

        async function scanDirFiles(fileName: string) {
          const fileAbsPath = dirPath + '/' + fileName;
          const scannedFile = await lstat(fileAbsPath);

          if (scannedFile.isSymbolicLink()) {
            return;
          }

          const newNode = new TreeNode(fileAbsPath);

          if (scannedFile.isDirectory()) {
            fileStack.push(newNode);
          }

          currentNode.appendChild(newNode);
        }

        await Promise.all(fileNameList.map(scanDirFiles));
      } catch (error) {
        if (!isNodeError(error)) throw error;
        if (!error.code) throw error;
        if (!allowedErrorCodes.has(error.code)) throw error;
      }
    }

    return rootNode;
};
